### ISSUE TITLE
Multi-GPU bf16 training on RTX PRO 6000 Blackwell (sm_120): resume training random lottery

### ISSUE THREAD

**Issue by philipshurpik** [pytorch/pytorch#193752]
### 🐛 Describe the bug

While training LORA adapters for Gemma 4 31B I encountered multiple times strange behaviour when starting and resuming training.
It occurred only with multiple GPUs (4 RTX PRO 6000), but everything was consistent and correct with one GPU training.

Initially was found while training translation adapters for different languages. But also reproducible with toy minimal examples. The bug reproduces on torch 2.11.0 (cu128 and cu130) **and torch 2.15.0.dev20260814+cu130 nightly**, with two separate trainer stacks (unsloth and plain transformers/TRL), so seems it is not trainer code.

## Environment

- 4x NVIDIA RTX PRO 6000 Blackwell Server Edition (96GB, sm_120), GCE g4 instance
- torch 2.11.0+cu128, torch 2.11.0+cu130, and torch 2.15.0.dev20260814+cu130 (all reproduce)
- Model: gemma-4-31b-it, bf16, LoRA r=8 via peft; 4-process DDP via torchrun
- transformers 5.5.0 / 5.13.1 / 5.16.0.dev0, trl 0.24-1.10 (all reproduce)

## How to reproduce (on 4x NVIDIA RTX PRO 6000)

1. Train any LoRA briefly and save a checkpoint (with optimizer state).
2. Relaunch and resume with `eval_on_start`, N times, with the identical command.
3. Compare the first eval loss across launches. Optionally run a probe — a forward pass on one
   fixed input tensor — on each rank (process) at each trainer startup stage.

I've created gist with toy NER LORA fine-tuning:
https://gist.github.com/philipshurpik/9fd62c0a085a9913b9870df38788bfe4

I've done train run for 10 steps. Initial eval loss was 0.448. And on saving checkpoint eval loss 0.1723.  
Six identical resume launches then printed eval_on_start = 6.501 | 0.1723 | 0.1884 | 6.501 | 0.2461 | 3.325  
Healthy eval draws at resume is the same as eval while checkpointing, bad draws are 1.1-40x off.

Single-GPU run (resume) is healthy (same eval loss 0.1723 for all 6 attempts):
    `CUDA_VISIBLE_DEVICES=0 python repro_lora.py`

## LORA finetuning start lottery (not only resume)
In my experience the same problem occurred also when starting training, also on 4 RTX PRO 6000 GPUs with same config. It was less frequent. And the problem of such starting fluctuations was not so obvious, but possibly it was the problem and it's limited not only to resume.

## Workaround found (and seems it can point on problem)
Running `trainer.evaluate()` before `train()` seems correctly initializes every rank in healthy state.

```
if resume:
    trainer._load_from_checkpoint(f'{out}/checkpoint-{steps}')
eval_loss = trainer.evaluate()['eval_loss']
print(f"*** Training with eval loss: {eval_loss}. Will be matched to eval_on_start loss of trainer.train***")

trainer.train(resume_from_checkpoint=f'{out}/checkpoint-{steps}' if resume else None)
```

This one seems settles all memory correctly and then training / resume is correct

Would be happy to provide extra guidance and help on reproducing if needed

### Versions

curl -sL https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py | python
Collecting environment information...
PyTorch version: 2.15.0.dev20260814+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-1022-gcp-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 1: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 2: NVIDIA RTX PRO 6000 Blackwell Server Edition
GPU 3: NVIDIA RTX PRO 6000 Blackwell Server Edition

Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  192
On-line CPU(s) list:                     0-191
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9B45
CPU family:                              26
Model:                                   2
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                5399.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext mwaitx ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 clzero xsaveerptr wbnoinvd arat avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b avx512_vp2intersect flush_l1d
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               4.5 MiB (96 instances)
L1i cache:                               3 MiB (96 instances)
L2 cache:                                96 MiB (96 instances)
L3 cache:                                384 MiB (12 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-191
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; IBPB on VMEXIT only
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.24.0.43
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.30.7
[pip3] nvidia-nvjitlink==13.4.46rc1
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.15.0.dev20260814+cu130
[pip3] triton==3.8.0+gitdf3f91dd
[conda] Could not collect

#####
# Because cuda installed from pip wheels and no nvidia driver version detected:
nvidia-smi
NVIDIA-SMI 610.43.02              KMD Version: 610.43.02     CUDA UMD Version: 13.3

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh
